src/renderingEngine.js

```diff
@@ -0,0 +1,887 @@
+    const {
+      img,
+      frames,
+      bitsAllocated,
+      pixelRepresentation,
+      thresholdValues,
+      color,
+      opacity,
+      columns,
+      rows
+    } = frameData
+
+    const signed = pixelRepresentation === 1 ? true : false;
+    if (frames) {
+      let {
+        pixelData, 
+        decodedframeInfo
+      } = this._checkImageTypeAndDecode(frames)
+      let bitsPerSample = decodedframeInfo.bitsPerSample;
+
+      if (!pixelData) {
+        // data downloaded uncompressed   
+        switch (bitsAllocated) {
+          case 8:
+            if (signed) {
+              pixelData = new Int8Array(frames)
+            } else {
+              pixelData = new Uint8Array(frames)
+            }
+            break;
+          case 16:
+            if (signed) {
+              pixelData = new Int16Array(frames)
+            } else {
+              pixelData = new Uint16Array(frames)
+            }
+            break;
+          default:
+            throw new Error(
+              'The pixel bit ' + bitsAllocated + 'is not supported by the offscreen render.'
+            );
+        }
+        bitsPerSample = bitsAllocated;
+      }
+      
+      // NOTE: we store the pixelData array and bitsPerSample in img, so we can apply again colorImageFrame
+      //       at the change of any blending parameter (opacity, color, clipping).     
+      img.pixelData = pixelData;
+      img.bitsPerSample = bitsPerSample;
+    }
+    
+    if (img.pixelData) {
+      // run offscreen render to color images
+      const renderedCanvas = this._render(
+        img.pixelData,
+        img.bitsPerSample, 
+        columns, 
+        rows, 
+        color, 
+        opacity, 
+        thresholdValues, 
+      );
+  
+      // econde back the image in png
+      img.src = renderedCanvas.toDataURL('image/png');
+  
+      // NOTE: ToBlob is async and provides smaller images,
+      // but here the renderingEngine is synch/sequential (one object, one gl context, etc..).
+      // When the OffscreenCanvas will be fully supported, 
+      // it can be used for a full web-workers async approach
+      // https://developer.mozilla.org/en-US/docs/Web/API/OffscreenCanvas
+
+      return true;
+    }
+
+    return false;
+  }
+
+  /** Creates image url from raw frames or rendered frame of RGB images. 
+   * Decodes the image if jp2/jpx (OpenJPEG) or jls (CharLS) and re-econde them into a png data url.
+   * If jpeg or png it just creates a url from a blob.
+   * The image type is automatically detected by checking the magic number
+   * https://en.wikipedia.org/wiki/Magic_number_(programming)#Magic_numbers_in_files 
+   * (see https://github.com/sindresorhus/file-type).
+   * 
+   * @param {number[]} frames - compressed image array
+   * @returns {boolean} blob.
+   * 
+   */
+  createURLFromRGBImage(frames) {
+    const  {
+      pixelData,
+      decodedframeInfo, 
+      mediaType
+    } = this._checkImageTypeAndDecode(frames, false);
```

Comment:
When testing locally, I am getting `this._checkImageTypeAndDecode(...) is undefined`? Did you test this with RGB images?

Reply:
yes I tested with 
const studyInstanceUID = '1.2.392.200140.2.1.1.1.2.799008771.2156.1519721160.421';
const seriesInstanceUID = '1.2.392.200140.2.1.1.1.3.799008771.2156.1519721160.422';
from https://server.dcmjs.org/dcm4chee-arc/aets/DCM4CHEE/rs

note that is the options retrieveRendered is true (as default) then for RGB images the fetching order is jp2, png, jpeg. In the case of the basic example the image then is retrieve as png For png _checkImageTypeAndDecode just return undefined and in createURLFromRGBImage I just create a blob.  I added inline codes more explicit descriptions

Reply:
I also just modfied the order as jp2, jpeg, png. since retrieving jpeg is faster, please let me know if ok